File: examples/cpp/example_06_managers.cpp
#include <chrono>
#include <filesystem>
#include <memory>
#include <string>

#include <glad/gl.h>

#include <renderer/engine/graphics/window_t.hpp>
#include <renderer/backend/graphics/opengl/resources_manager_t.hpp>
#include <renderer/engine/input_manager_t.hpp>
#include <renderer/backend/graphics/opengl/vertex_buffer_opengl_t.hpp>
#include <renderer/backend/graphics/opengl/index_buffer_opengl_t.hpp>
#include <renderer/backend/graphics/opengl/vertex_array_opengl_t.hpp>

#include <utils/logging.hpp>

#if defined(RENDERER_IMGUI)
#include <imgui.h>
#endif  // RENDERER_IMGUI

struct Engine {
    ::renderer::Window::ptr window{nullptr};
    ::renderer::opengl::ResourcesManager::ptr resources_manager{nullptr};
    ::renderer::InputManager::ptr input_manager{nullptr};
    std::chrono::steady_clock::time_point stamp_now{};
    std::chrono::steady_clock::time_point stamp_bef{};
    double time = 0.0;
    double delta = 0.0;
};

// NOLINTNEXTLINE
Engine g_engine;

auto main() -> int {
    ::renderer::WindowConfig win_config;
    win_config.backend = ::renderer::eWindowBackend::TYPE_GLFW;
    win_config.width = 1024;
    win_config.height = 768;
    win_config.title = "Example 06 - Managers";
    win_config.gl_version_major = 3;
    win_config.gl_version_minor = 3;

    g_engine.window = ::renderer::Window::CreateWindow(win_config);
    g_engine.window->RegisterKeyboardCallback([&](int key, int action, int) {
        g_engine.input_manager->CallbackKey(key, action);
        if (key == ::renderer::keys::KEY_ESCAPE) {
            g_engine.window->RequestClose();
        }
    });
    g_engine.window->RegisterMouseButtonCallback(
        [&](int button, int action, double x, double y) {
            g_engine.input_manager->CallbackMouseButton(
                button, action, static_cast<float>(x), static_cast<float>(y));
        });
    g_engine.window->RegisterMouseMoveCallback([&](double x, double y) {
        g_engine.input_manager->CallbackMouseMove(static_cast<float>(x),
                                                  static_cast<float>(y));
    });
    g_engine.window->RegisterScrollCallback([&](double xOff, double yOff) {
        g_engine.input_manager->CallbackScroll(static_cast<float>(xOff),
                                               static_cast<float>(yOff));
    });

    g_engine.resources_manager =
        std::make_shared<::renderer::opengl::ResourcesManager>();
    g_engine.input_manager = std::make_shared<::renderer::InputManager>();

    const auto SHADERS_FOLDER =
        std::filesystem::path(__FILE__).parent_path().parent_path() /
        "resources" / "shaders";
    const auto VERT_SHADER_FILE = SHADERS_FOLDER / "basic2d_vert.glsl";
    const auto FRAG_SHADER_FILE = SHADERS_FOLDER / "basic2d_frag.glsl";
    auto program = g_engine.resources_manager->LoadProgram(
        "basic2d", VERT_SHADER_FILE.string(), FRAG_SHADER_FILE.string());
    program->Build();
    if (!program->IsValid()) {
        LOG_CORE_ERROR("There was an error building the shader program");
        return 1;
    }

    const auto IMAGES_FOLDER =
        std::filesystem::path(__FILE__).parent_path().parent_path() /
        "resources" / "images";
    const auto TEXTURE_FILE = IMAGES_FOLDER / "container.jpg";
    auto texture = g_engine.resources_manager->LoadTexture(
        "container", TEXTURE_FILE.string());

    // -------------------------------------------------------------------------
    // Setup the primitive to be drawn

    // clang-format off
    // NOLINTNEXTLINE
    float buffer_data[] = {
        /*|   pos   |  texture  |*/
        -0.5F, -0.5F, 0.0F, 0.0F, // NOLINT
         0.5F, -0.5F, 2.0F, 0.0F, // NOLINT
         0.5F,  0.5F, 2.0F, 2.0F, // NOLINT
        -0.5F,  0.5F, 0.0F, 2.0F // NOLINT
    };

    // NOLINTNEXTLINE
    uint32_t buffer_indices[] = {
        0, 1, 2,  // first triangle
        0, 2, 3  // second triangle
    };

    constexpr uint32_t NUM_VERTICES = 6;
    // clang-format on

    ::renderer::opengl::OpenGLBufferLayout layout = {
        {"position", ::renderer::eElementType::FLOAT_2, false},
        {"texcoord", ::renderer::eElementType::FLOAT_2, false}};

    auto vbo = std::make_shared<::renderer::opengl::OpenGLVertexBuffer>(
        layout, renderer::eBufferUsage::STATIC,
        static_cast<uint32_t>(sizeof(buffer_data)), buffer_data);
    auto ibo = std::make_shared<::renderer::opengl::OpenGLIndexBuffer>(
        renderer::eBufferUsage::STATIC, NUM_VERTICES, buffer_indices);

    auto vao = std::make_shared<::renderer::opengl::OpenGLVertexArray>();
    vao->AddVertexBuffer(vbo);
    vao->SetIndexBuffer(ibo);
    // -------------------------------------------------------------------------

    g_engine.stamp_bef = std::chrono::steady_clock::now();
    g_engine.stamp_now = std::chrono::steady_clock::now();

    while (g_engine.window->active()) {
        g_engine.window->Begin();

        if (g_engine.input_manager->IsKeyDown(::renderer::keys::KEY_ESCAPE)) {
            g_engine.window->RequestClose();
        }

        auto cursor_position = g_engine.input_manager->GetCursorPosition();
        LOG_INFO("Cursorposition: ({0},{1})", cursor_position.x(),
                 cursor_position.y());
        LOG_INFO("MouseButtonLeft: {0}", g_engine.input_manager->IsMouseDown(
                                             renderer::mouse::BUTTON_LEFT));
        LOG_INFO("MouseButtonMiddle: {0}", g_engine.input_manager->IsMouseDown(
                                               renderer::mouse::BUTTON_MIDDLE));
        LOG_INFO("MouseButtonRight: {0}", g_engine.input_manager->IsMouseDown(
                                              renderer::mouse::BUTTON_RIGHT));

        // Draw the cube
        {
            program->Bind();
            texture->Bind();
            vao->Bind();

            glDrawElements(GL_TRIANGLES, NUM_VERTICES, GL_UNSIGNED_INT,
                           nullptr);

            vao->Unbind();
            texture->Unbind();
            program->Unbind();
        }
#if defined(RENDERER_IMGUI)
        ImGui::Begin("Demo options");
        {
            // NOLINTNEXTLINE
            ImGui::Text("time: %.5f, delta: %.5f", g_engine.time,
                        g_engine.delta);
        }
#endif  // RENDERER_IMGUI

        g_engine.window->End();

        // Keep track of time
        {
            g_engine.stamp_now = std::chrono::steady_clock::now();
            g_engine.delta =
                std::chrono::duration_cast<std::chrono::duration<double>>(
                    g_engine.stamp_now - g_engine.stamp_bef)
                    .count();
            g_engine.time += g_engine.delta;
            g_engine.stamp_bef = g_engine.stamp_now;
        }
    }

    return 0;
}

File: examples/python/example_06_managers.py
#!/usr/bin/env python

from pathlib import Path
from typing import Optional

import numpy as np
from OpenGL.GL import *  # type: ignore

import renderer as rdr
from renderer.opengl import (
    OpenGLBufferElement,
    OpenGLBufferLayout,
    OpenGLProgram,
    OpenGLVertexArray,
    OpenGLVertexBuffer,
    OpenGLIndexBuffer,
    ResourcesManager,
)


class Engine:
    def __init__(self):
        self.window: Optional[rdr.Window] = None
        self.resources_manager: Optional[ResourcesManager] = None
        self.input_manager: Optional[rdr.InputManager] = None


g_engine = Engine()


def main() -> int:
    global g_engine
    win_config = rdr.WindowConfig()
    win_config.backend = rdr.WindowBackend.TYPE_GLFW
    win_config.width = 1024
    win_config.height = 768
    win_config.title = "Example 06 - Managers"
    win_config.gl_version_major = 3
    win_config.gl_version_minor = 3

    g_engine.window = rdr.Window.CreateWindow(win_config)
    assert g_engine.window is not None

    def keyboard_callback(key: int, action: int, _: int) -> None:
        if g_engine.input_manager is not None:
            g_engine.input_manager.CallbackKey(key, action)

    g_engine.window.RegisterKeyboardCallback(keyboard_callback)

    def mouse_button_callback(
        button: int, action: int, x: float, y: float
    ) -> None:
        if g_engine.input_manager is not None:
            g_engine.input_manager.CallbackMouseButton(button, action, x, y)

    g_engine.window.RegisterMouseButtonCallback(mouse_button_callback)

    def mouse_move_callback(x: float, y: float) -> None:
        if g_engine.input_manager is not None:
            g_engine.input_manager.CallbackMouseMove(x, y)

    g_engine.window.RegisterMouseMoveCallback(mouse_move_callback)

    def scroll_callback(xOff: float, yOff: float) -> None:
        if g_engine.input_manager is not None:
            g_engine.input_manager.CallbackScroll(xOff, yOff)

    g_engine.window.RegisterScrollCallback(scroll_callback)

    g_engine.resources_manager = ResourcesManager()
    g_engine.input_manager = rdr.InputManager()

    assert g_engine.resources_manager is not None
    assert g_engine.input_manager is not None

    SHADERS_FOLDER = Path(__file__).parent.parent / "resources" / "shaders"
    VERT_SHADER_FILE = SHADERS_FOLDER / "basic2d_vert.glsl"
    FRAG_SHADER_FILE = SHADERS_FOLDER / "basic2d_frag.glsl"

    program = g_engine.resources_manager.LoadProgram(
        "basic2d",
        str(VERT_SHADER_FILE.resolve()),
        str(FRAG_SHADER_FILE.resolve()),
    )
    program.Build()
    if not program.valid:
        print("There was an error building the shader program")
        return 1

    IMAGES_FOLDER = Path(__file__).parent.parent / "resources" / "images"
    TEXTURE_FILE = IMAGES_FOLDER / "container.jpg"
    texture = g_engine.resources_manager.LoadTexture(
        "container", str(TEXTURE_FILE.resolve())
    )

    # fmt: off
    buffer_data = np.array([
        # |--pos--|-texture-|#
        -0.5, -0.5, 0.0, 0.0,  # noqa: E131
         0.5, -0.5, 2.0, 0.0,  # noqa: E131
         0.5,  0.5, 2.0, 2.0,  # noqa: E131
        -0.5,  0.5, 0.0, 2.0,  # noqa: E131
    ], dtype=np.float32)

    NUM_VERTICES = 6

    buffer_indices = np.array([
        0, 1, 2,
        0, 2, 3
    ], dtype=np.uint32)
    # fmt: on

    layout = OpenGLBufferLayout()
    layout.AddElement(
        OpenGLBufferElement("position", rdr.ElementType.FLOAT_2, False)
    )
    layout.AddElement(
        OpenGLBufferElement("texcoord", rdr.ElementType.FLOAT_2, False)
    )

    vbo = OpenGLVertexBuffer(layout, rdr.BufferUsage.STATIC, buffer_data)
    ibo = OpenGLIndexBuffer(rdr.BufferUsage.STATIC, buffer_indices)

    vao = OpenGLVertexArray()
    vao.AddVertexBuffer(vbo)
    vao.SetIndexBuffer(ibo)

    while g_engine.window.active:
        g_engine.window.Begin()

        if g_engine.input_manager.IsKeyDown(rdr.Keys.KEY_ESCAPE):
            g_engine.window.RequestClose()

        cursor_position = g_engine.input_manager.GetCursorPosition()
        print(f"CursorPosition: ({cursor_position.x}, {cursor_position.y})")
        print(
            "MouseButtonLeft: {}".format(
                g_engine.input_manager.IsMouseDown(rdr.Mouse.BUTTON_LEFT)
            )
        )
        print(
            "MouseButtonMiddle: {}".format(
                g_engine.input_manager.IsMouseDown(rdr.Mouse.BUTTON_MIDDLE)
            )
        )
        print(
            "MouseButtonRight: {}".format(
                g_engine.input_manager.IsMouseDown(rdr.Mouse.BUTTON_RIGHT)
            )
        )

        program.Bind()
        texture.Bind()
        vao.Bind()

        glDrawElements(GL_TRIANGLES, NUM_VERTICES, GL_UNSIGNED_INT, None)

        vao.Unbind()
        texture.Unbind()
        program.Unbind()

        g_engine.window.End()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
